### PR TITLE
perf(rocm): fuse GDN in_proj_qkvz + in_proj_ba into single GEMM

### PR DESCRIPTION
**Pull request by Xu-Sheng-lin**
In Qwen3-Next GDN, in_proj_qkvz and in_proj_ba are two separate column-parallel linears that take the SAME hidden_states input and produce non-overlapping output ranges. Fusing them into a single in_proj_fused GEMM (cat along dim=1 of the weights) reads hidden_states only once and amortizes the GEMM launch.

The fusion is a pure compute-graph rearrangement:
  out_qkvz = hidden_states @ qkvz_w        (N=6144 on Qwen3.5-9B TP=2)
  out_ba   = hidden_states @ ba_w          (N=32)
becomes:
  fused_out = hidden_states @ cat([qkvz_w, ba_w], dim=1)  (N=6176)
  out_qkvz  = fused_out[..., :6144]
  out_ba    = fused_out[..., 6144:]

Decode (M=1) is HBM-bandwidth bound, so reading the hidden tensor once for both projections matters. Prefill (M=15384) is compute-bound, so the fusion GEMM is mildly slower than the sum of the two (N=6176 picks a slightly less optimal hipBLASLt tile vs N=6144's perfect MT192x32).

Measured on Qwen3.5-9B TP=2, 15K prefill + 100 decode (n=10, unique prompts):
  - TTFT:  1036.37 ms -> 1035.06 ms   (-1.31 ms, noise)
  - TPOT:    6.992 ms ->   6.775 ms   (-0.217 ms, -3.1%)
  - Total:  1728.6  ms -> 1705.8  ms  (-22.8 ms / request)

Decode-bound workloads (long generation) get most of the win; short-output workloads see ~0 net change. Independent of (and composes with) the scatter_qkv prefill optimization on feat/gdn-scatter-qkv-triton.

FP8/quantized fallback: qkvz has scales but ba doesn't, dtype mismatch makes direct cat impossible. The fallback keeps the original 2-GEMM path.

### CODE REVIEW

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/2 · P3/0

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **融合权重额外常驻一份大矩阵** @ `rtp_llm/models_py/model_desc/qwen3_next.py:532`
  - 建议：将融合下沉到权重加载阶段，或在模块初始化后释放/替换原 qkvz、ba 引用，并补充显存开销评估。
- **缺少融合分支等价性测试** @ `rtp_llm/models_py/model_desc/qwen3_next.py:526`
  - 建议：增加小尺寸 Qwen3NextGatedDeltaNet 单测，固定权重比较 fused/非 fused 输出，并验证带 `linear_attn_qkvz_s` 时走原路径。

### Checklist Violations (5 fail / 88 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `缺少融合分支等价性测试`
  _新增 fused/fallback 分支但 PR 没有新增等价性测试。_
- [6.1] Tests — 边界 case 覆盖（空、单元素、最大值）  → issue `缺少融合分支等价性测试`
  _缺少小 batch/decode M=1 与 quantized fallback 的边界覆盖。_
- [6.1] Tests — 分布式/跨平台变更有对应覆盖  → issue `缺少融合分支等价性测试`
  _融合路径影响 TP 下每层 projection，未见 TP>1/ROCm 等对应回归覆盖。_

**RTP-LLM Checklist**
- [B] 正确性与逻辑 — 边界 case（空输入、单元素、最大值）  → issue `缺少融合分支等价性测试`
  _新增 fused 路径未补充 M=1 decode 或 fallback 边界测试。_
- [D] 性能 — 通信 / buffer 开销评估  → issue `融合权重额外常驻一份大矩阵`
  _`torch.cat` 构造 fused_w 后原 qkvz/ba 权重仍被模型权重对象持有，缺少显存开销处理。_

### Strengths
- 融合实现复用 `LinearFactory.create_linear`，没有新增平台专用调用分支。
- 量化路径保留原 two-GEMM fallback，降低 FP8 路径被误改的风险。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/1 · P3/0

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **融合分支置空投影模块会破坏现有 CP 测试** @ `rtp_llm/models_py/model_desc/qwen3_next.py:545`
  - 建议：保留一个统一的 projection helper 供 forward 和测试复用，或同步更新 CP 测试走 fused 分支的切片逻辑。

### Checklist Violations (1 fail / 56 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `融合分支置空投影模块会破坏现有 CP 测试`
  _新增测试没有覆盖已有 CP 测试依赖的 module.in_proj_qkvz/module.in_proj_ba 属性，融合分支将其置空会破坏现有测试。_

### Strengths
- 改动将 BF16 qkvz/ba 两次 GEMM 合并为一次，并保留了带 qkvz scale 的量化路径回退。
- 新增测试覆盖了 BF16 融合开关、量化 scale 分支派发和源权重释放行为。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/1 · P3/0

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **量化回退测试只复刻条件，未覆盖构造分支** @ `rtp_llm/models_py/model_desc/test/qwen3_next_qkvz_ba_fusion_test.py:128`
  - 建议：用带 scale 的最小权重实例化模块，或 mock LinearFactory，直接断言 _qkvz_ba_fused=False、in_proj_fused=None 且两个原始 projection 被创建。

### Checklist Violations (2 fail / 92 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `量化回退测试只复刻条件，未覆盖构造分支`
  _新增单测覆盖 BF16 融合，但量化 fallback 测试只断言 dict.get 条件，未实例化模块验证 fallback 构造。_

**RTP-LLM Checklist**
- [H] 测试与 CI — 测试覆盖充分：大重构等价覆盖，新功能端到端测试  → issue `量化回退测试只复刻条件，未覆盖构造分支`
  _BF16 融合有等价测试，但量化 fallback 只测试 dict.get 条件，没有端到端实例化覆盖。_

### Strengths
- BF16 融合路径把投影调用收敛到 _input_project，主 forward 和 CP 测试共用同一入口，降低了后续分支漂移风险。
- 新增测试覆盖了融合数学等价、BF16 分支开关和权重释放行为，对主要优化目标有直接保护。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** BLOCKING

**Summary:** P0/0 · P1/1 · P2/0 · P3/0

### Blocking Issues

#### P1
- **融合后原始权重 key 被删除，在线权重更新失效** @ `rtp_llm/models_py/model_desc/qwen3_next.py:540`
  - 建议：为 fused 权重接入原始 qkvz/ba 的更新映射，或保留可更新的原始 tensor 并在更新后同步重建/刷新 `in_proj_fused`。

### Checklist Violations (3 fail / 56 total)

**General Principles Checklist**
- [6.1] Architecture — 状态不变量：创建/更新/失败/重试/回滚路径有效  → issue `融合后原始权重 key 被删除，在线权重更新失效`
  _构造期删除原始 qkvz/ba key 后，`ModelWeights` 中不再保留与在线更新同名的权重状态。_
- [6.1] Architecture — 兼容性：公开 API/持久数据/配置/环境迁移安全  → issue `融合后原始权重 key 被删除，在线权重更新失效`
  _外部仍按原始权重名更新 qkvz/ba，但 BF16 构造后模型权重字典已经移除这些 key。_
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `融合后原始权重 key 被删除，在线权重更新失效`
  _新增测试覆盖构造、shape 和 FP8 fallback，但没有覆盖 update_weights/ModelWeights 更新 fused 权重后的行为。_

### Strengths
- 新增测试覆盖了 BF16 融合分支、FP8 fallback 分支，并让 CP 线性注意力测试复用 `_input_project`。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** BLOCKING

**Summary:** P0/0 · P1/1 · P2/0 · P3/0

### Blocking Issues

#### P1
- **新增融合测试按错误权重布局切片导致失败** @ `rtp_llm/models_py/model_desc/test/qwen3_next_qkvz_ba_fusion_test.py:255`
  - 建议：按真实布局校验：CUDA 路径用 in_proj_fused.weight[:_qkvz_size, :].T 和 in_proj_fused.weight[_qkvz_size:, :].T，或直接校验 weights dict 视图与底层 storage/update 后 forward 输出。

### Checklist Violations (2 fail / 92 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `新增融合测试按错误权重布局切片导致失败`
  _新增聚焦单测存在布局断言错误，H20 目标会失败，不能作为有效覆盖。_

**RTP-LLM Checklist**
- [H] 测试与 CI — 测试覆盖充分：大重构等价覆盖，新功能端到端测试  → issue `新增融合测试按错误权重布局切片导致失败`
  _新增测试覆盖方向正确，但在线更新断言按错误 layout 切片，导致测试本身失败。_

### Strengths
- BF16 融合路径和 FP8 fallback 路径都补了针对性测试，覆盖了主要分支。
- 实现把在线权重更新场景纳入设计，避免融合后丢失原有 update_layer_weight 语义。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/0 · P3/0

lgtm ready to ci

### Checklist ✅ (56 items passed)

### Strengths
- 融合路径保留量化 fallback，并用 _input_project 统一 forward 与 CP 测试调用。
- 新增测试覆盖权重视图与在线更新语义，能防止 fused buffer 与 ModelWeights 脱节。

**Review by LLLLKKKK — APPROVED** [review]

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1019

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/1 · P3/0

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **ROCm/swizzleA 目标路径缺少测试覆盖** @ `rtp_llm/models_py/model_desc/test/BUILD:8`
  - 建议：补充 ROCm/MI308X 测试目标或参数化用例，覆盖 no-swizzle 与 use_swizzleA/bpreshuffle 分支。

### Checklist Violations (3 fail / 92 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `ROCm/swizzleA 目标路径缺少测试覆盖`
  _已有 H20 聚焦单测，但缺少 ROCm/swizzleA 相关覆盖。_
- [6.1] Tests — 分布式/跨平台变更有对应覆盖  → issue `ROCm/swizzleA 目标路径缺少测试覆盖`
  _新增 BUILD 只配置 H20，未覆盖 PR 目标 ROCm/swizzleA 路径。_

**RTP-LLM Checklist**
- [H] 测试与 CI — 测试覆盖充分：大重构等价覆盖，新功能端到端测试  → issue `ROCm/swizzleA 目标路径缺少测试覆盖`
  _新增测试覆盖 CUDA/H20，未覆盖 ROCm/use_swizzleA Linear 分支。_

### Strengths
- 新增测试覆盖 BF16 fusion、量化 fallback、权重 view 与在线权重更新语义。
- forward 和 CP 测试统一复用 _input_project，降低融合/非融合路径漂移风险。
